Implement 'resize set width|height <ppt> ppt' command

livanh opened the pull request:
This patch allows resizing tiling containers to exactly the given size (in ppt). The sintax is as in the title. The patch includes only the code now, I will update documentation if needed before merging.

Commit:
Implement 'resize set <width> ppt <height> ppt' command for tiling windows

Airblader commented on src/commands.c:
```diff
@@ -633,6 +633,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+ */
+void cmd_resize_tiling_set(I3_CMD, const char *direction, long ppt) {
+    DLOG("set size to %ld ppt in %s direction\n", ppt, direction);
+    if (ppt<5){
```
Why this arbitrary restriction?

livanh replied:
I saw that cmd_resize_tiling_width_height() returns without resizing if any of the involved windows could become smaller than 5%. I assumed that means "we don't want any window to be smaller than 5%", but feel free to suggest a different value. At the very least, we should restrict ppt to be non-negative.

Airblader commented on src/commands.c:
```diff
@@ -633,6 +633,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+void cmd_resize_tiling_set(I3_CMD, const char *direction, long ppt) {
+    DLOG("set size to %ld ppt in %s direction\n", ppt, direction);
+    if (ppt<5){
+        ELOG("Resize failed: requested size is too small (%ld)\n", ppt);
```
This (and all other) log messages containing a value in ppt should include the `%` sign to provide a better understanding.

livanh replied:
Ok, I'll do so

Airblader commented on src/commands.c:
```diff
@@ -633,6 +633,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+
+        double current_percent = current->con->percent;
+        if (current_percent > ppt) {
+             if (!cmd_resize_tiling_width_height(current_match, cmd_output,
```
Since we're doing the same call in both cases, it'd IMHO be better to use the if/else to define the arguments and only have one path actually execute the call.

Airblader replied:
In fact, I don't really like that we are calling the handler of another command here, and especially like this. That feels quite hacky. This logic should be handled correctly in `con_*` functions, but will require some refactoring, I think.

livanh replied:
No problem, I'll do it this way.

Airblader commented on src/commands.c:
```diff
@@ -633,6 +633,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+        if (current_percent > ppt) {
+             if (!cmd_resize_tiling_width_height(current_match, cmd_output,
+                                                    current->con, "shrink", direction, (int)(current_percent*100)-ppt))
+                    return;
```
Shouldn't this be logged? Also, in this case we are not returning any response on the IPC interface, which is bad.

livanh replied:
This is how it's done in cmd_resize(). In fact, cmd_resize_tiling_width_height() prints some logging messages on its own, and using this command with i3-msg returns [{"success":true}] (or false). Is this the right way to check the response?

Airblader reviewed (commented):
I only have a few quick comments, but this should not be taken as a merge decision.

We do have https://github.com/i3/i3/issues/2608 for that discussion which hasn't really reached any decision yet.

livanh commented:
I did't notice the discussion in #2608. If I understand it correctly, this would be a partial solution to that issue. At this time I'm not interested in doing the remaining part (i.e. associating some weight to each window), but it is an interesting idea nonetheless.

Airblader commented:
For the record, I don't think we'd implement something like the proposed weights idea. If at all, I think we'd stick with something like this.

livanh commented:
I just tried and it works also with `for_window` matching, as asked in #2608. Let me know if you decide to merge and what changes are needed.

pevu commented:
Please merge, I'd like to have this as well. Right now I have to write hacky wrapper scripts to get the same functionality, as I've seen other people do too.

Airblader commented on src/commands.c:
```diff
@@ -632,6 +632,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+ */
+void cmd_resize_tiling_set(I3_CMD, const char *direction, long ppt) {
+    DLOG("set size to %ld ppt in %s direction\n", ppt, direction);
+    if (ppt < 5) {
```
Let's please just remove this check here.

Airblader commented on src/commands.c:
```diff
@@ -632,6 +632,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+
+        double current_percent = current->con->percent;
+        if (current_percent > ppt) {
+            if (!cmd_resize_tiling_width_height(current_match, cmd_output,
```
Can we just slightly change this logic to instead execute the resize for all matched containers regardless if it failed for one of them? We just collect the accumulated result of all calls such that in the end we report a success if and only if all were successful and a failure if a single one failed, but that way the resize still worked for all matches where it could work.

Airblader commented on src/commands.c:
```diff
@@ -632,6 +632,47 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+    }
+
+    cmd_output->needs_tree_render = true;
+    // XXX: default reply for now, make this a better reply
```
Please remove this comment (I know it's in other places, but we don't need it here)

Airblader commented:
I'm willing to merge this feature, however there's a few things:

1. I'd prefer the syntax `resize set 35 ppt 50 ppt` which is in line with the current `px` symbol (which is optional). Otherwise we have two different syntax variations for doing the same thing.
2. The userguide has to be updates (`docs/userguide`).
3. This is easily testable, so we should have some tests for this.
4. I've added a few review comments which should be addressed (as much as they make sense after the other points above).

pevu commented:
But what if you only wanted to adjust the height of the container? I thought this is for tiling containers, where usually either the height or the width is changed, not both at the same time.

Airblader commented:
Point taken. I still feel it'd be good to have the option to change both in an atomic operation. Perhaps we could use a placeholder value representing a nop. @stapelberg What do you think?

livanh commented:
My intended use is the same as @pevu, i.e. only resize in one direction. I can change the code to use the syntax you proposed (with a nop placeholder), but I'd like to clarify two points first:

1.  What happens if the unit (px/ppt) is omitted? Should we assume px for backwards compatibility, or turn it into an error?
2. Is it ok to call `cmd_resize_tiling_width_height` internally, or is it better to make a dedicated function?

Airblader commented:
> with a nop placeholder

Please note that we should first discuss what that placeholder looks like and if that's really what we want to go for. Hence me tagging @stapelberg to discuss it. :-)

> What happens if the unit (px/ppt) is omitted? Should we assume px for backwards compatibility, or turn it into an error?

We don't like breaking things, so we would be falling back to px here.

> Is it ok to call cmd_resize_tiling_width_height internally, or is it better to make a dedicated function?

I actually think calling it is OK.

stapelberg commented:
I agree with @Airblader: setting width/height in the same command seems reasonable to me.

livanh commented:
I think I did everything, except for the testcase. That would require some study and experimentation, which I do not have time to do today.

 As for the placeholder, I simply assumed a value of zero or less to mean "nop" for the corresponding direction, as they would not make sense anyway.

However, I just noticed that there is a conflict with #3023 (already merged), where "set <ppt> ppt <ppt> ppt" means to resize a floating container in percentage points. A possible way out would be to move my logic into `cmd_resize_set`, and follow the path for floating or tiling containers depending on the currently focused container. Does it make sense?

stapelberg commented on docs/userguide:
```diff
@@ -2311,8 +2312,11 @@ space from all the other containers. The optional pixel argument specifies by
+default is 10 percentage points).
+
+Note that +resize set+ takes the first form for floating containers and the
+second one for tiling containers. In the latest case, a negative or zero value
```
s/latest/latter/

stapelberg commented on include/commands.h:
```diff
@@ -74,6 +74,12 @@ void cmd_resize_set(I3_CMD, long cwidth, long cheight);
  */
 void cmd_resize(I3_CMD, const char *way, const char *direction, long resize_px, long resize_ppt);
 
+/*
```
nit: start function comments in header files with `/**`

stapelberg commented on src/commands.c:
```diff
@@ -632,6 +632,99 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+                   (con_orientation(target->parent) != HORIZ || con_num_children(target->parent) == 1))
+                target = target->parent;
+
+            /* Calculate new size for th target container */
```
nit: s/th/the/

stapelberg commented on src/commands.c:
```diff
@@ -632,6 +632,99 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+
+    owindow *current;
+    bool success = true;
+    char *action_grow_string = "grow";
```
Why declare these as a variable? You can just use "grow" below.

stapelberg commented on src/commands.c:
```diff
@@ -632,6 +632,99 @@ static bool cmd_resize_tiling_width_height(I3_CMD, Con *current, const char *way
+        }
+
+        if (cwidth > 0) {
+            Con *target = current->con;
```
Can `resize_find_tiling_participants` not be used here? It looks very similar to this block of code.

livanh commented:
I rebased over the latest `next` branch, moved the code into `cmd_resize_set`, removed the old parts, and added a test case. I can also squash the commits for a cleaner history, if you want.

For the comment on `resize_find_tiling_participants`, I tried using it, but it's not exactly the same. Sometimes the command works as expected, and sometimes it does nothing.
By looking at the code, the main difference seems to be in the `con_num_children(target->parent) == 1` condition. I actually copied this code from somewhere else in the file. If we can decide that one variant is definitely better than the other, I can handle this as a separate refactoring task.

Airblader commented on src/commands.c:
```diff
@@ -681,33 +681,104 @@ void cmd_resize(I3_CMD, const char *way, const char *direction, long resize_px,
         if ((floating_con = con_inside_floating(current->con))) {
             Con *output = con_get_output(floating_con);
-            if (mode_width && strcmp(mode_width, "ppt") == 0) {
+            if (cwidth == 0)
```
Please use `{}` braces everywhere.

Airblader commented on src/commands.c:
```diff
@@ -681,33 +681,104 @@ void cmd_resize(I3_CMD, const char *way, const char *direction, long resize_px,
+            if (cwidth > 0 && mode_width && strcmp(mode_width, "ppt") == 0) {
+                Con *target = current->con;
+                /* get the appropriate current container (skip stacked/tabbed cons) */
+                while (target->parent->layout == L_STACKED ||
```
Please use `{}` braces for loops as well. Also, don't these two conditions fit on one line?

stapelberg commented:
> For the comment on resize_find_tiling_participants, I tried using it, but it's not exactly the same. Sometimes the command works as expected, and sometimes it does nothing.
> By looking at the code, the main difference seems to be in the con_num_children(target->parent) == 1 condition. I actually copied this code from somewhere else in the file.

So there is another duplicate of this functionality? Can you point me to it please?

> If we can decide that one variant is definitely better than the other, I can handle this as a separate refactoring task.

I’m a bit worried about pushing it to a later point in time, because then we might easily forget about it :).

Naively, I would argue that a variant which works in all the cases is better than the others. Could you please look into sending a PR which reduces duplication, so that we can make use of the new function in this PR?

Thanks!

livanh commented:
Please see #3111

stapelberg commented:
Could you clean up the git history of this PR please? See git’s “interactive rebase” feature. In the end, we’d like to have one git commit per logical change. Changes based on the code review don’t need to be reflected in the history. Thanks!

livanh commented:
I made everything into one commit, since we don't really need the initial syntax anymore. Is that ok?

stapelberg commented:
Yep. Thanks!
